endpoints/events/vtrack.go

```diff
@@ -0,0 +1,318 @@
+		return req, err
+	}
+
+	if err := json.Unmarshal(requestJson, req); err != nil {
+		return req, err
+	}
+
+	for _, bcr := range req.Puts {
+		if bcr.BidID == "" {
+			err = error(&errortypes.BadInput{Message: fmt.Sprint("'bidid' is required field and can't be empty")})
+			return req, err
+		}
+
+		if bcr.Bidder == "" {
+			err = error(&errortypes.BadInput{Message: fmt.Sprint("'bidder' is required field and can't be empty")})
+			return req, err
+		}
+	}
+
+	return req, nil
+}
+
+// handleVTrackRequest handles a VTrack request
+func (v *vtrackEndpoint) handleVTrackRequest(ctx context.Context, req *BidCacheRequest, account *config.Account) (*BidCacheResponse, []error) {
+	var biddersAllowingVastUpdate []string
+
+	if account.EventsEnabled {
+		biddersAllowingVastUpdate = getBiddersAllowingVastUpdate(req, &v.BidderInfos, v.Cfg.VTrack.AllowUnknownBidder)
+	}
+
+	// cache data
+	r, errs := v.cachePutObjects(ctx, req, biddersAllowingVastUpdate, account.ID)
+
+	// handle pbs caching errors
+	if len(errs) != 0 {
+		glog.Errorf("Error(s) updating vast: %v", errs)
+		return nil, errs
+	}
+
+	// build response
+	response := &BidCacheResponse{
+		Responses: []CacheObject{},
+	}
+
+	for _, uuid := range r {
+		response.Responses = append(response.Responses, CacheObject{
+			UUID: uuid,
+		})
+	}
+
+	return response, nil
+}
+
+// cachePutObjects caches BidCacheRequest data
+func (v *vtrackEndpoint) cachePutObjects(ctx context.Context, req *BidCacheRequest, biddersAllowingVastUpdate []string, accountId string) ([]string, []error) {
+	var cacheables []prebid_cache_client.Cacheable
+
+	if len(biddersAllowingVastUpdate) > 0 {
+		sort.Strings(biddersAllowingVastUpdate)
+	}
+
+	for _, c := range req.Puts {
+
+		nc := &prebid_cache_client.Cacheable{
+			Type:       c.Type,
+			Data:       c.Data,
+			TTLSeconds: c.TTLSeconds,
+			Key:        c.Key,
+		}
+
+		if contains(biddersAllowingVastUpdate, c.Bidder) && nc.Data != nil {
+			nc.Data = modifyVastXml(v.Cfg.ExternalURL, nc.Data, c.BidID, c.Bidder, accountId, c.Timestamp)
+		}
+
+		cacheables = append(cacheables, *nc)
+	}
+
+	return v.Cache.PutJson(ctx, cacheables)
+}
+
+// getBiddersAllowingVastUpdate returns a list of bidders that allow VAST XML modification
+func getBiddersAllowingVastUpdate(req *BidCacheRequest, bidderInfos *adapters.BidderInfos, allowUnknownBidder bool) []string {
+	var bl = make(map[string]struct{})
+
+	for _, bcr := range req.Puts {
+		if _, ok := bl[bcr.Bidder]; isAllowVastForBidder(&bcr, bidderInfos, allowUnknownBidder) && !ok {
+			bl[bcr.Bidder] = struct{}{}
+		}
+	}
+
+	length := len(bl)
+	if length == 0 {
+		return []string{}
+	}
+
+	keys := make([]string, 0, length)
+	for key := range bl {
+		keys = append(keys, key)
+	}
+
+	return keys
+}
+
+// isAllowVastForBidder checks if a bidder is active and allowed to modify vast xml data
+func isAllowVastForBidder(r *prebid_cache_client.Cacheable, bidderInfos *adapters.BidderInfos, allowUnknownBidder bool) bool {
```

Comment:
nitpick: if the only field from `r` that this function uses is `r.Bidder`, it could receive that as a parameter instead of the entire structure.